File: boxroom/jprelude-webui-java/src/main/java/org/jprelude/webui/data/PagingPosition.java
package org.jprelude.webui.data;

public class PagingPosition {
	private final int pageIdx;
	private final int pageSize;
	private final int totalItemCount;

	public PagingPosition(
			final int pageIdx,
			final int pageSize,
			final int totalItemCount) {

		this.pageIdx = pageIdx;
		this.pageSize = pageSize;
		this.totalItemCount = totalItemCount;
	}

	public int getPageIdx() {
		return this.pageIdx;
	}

	public int getPageSize() {
		return this.pageSize;
	}

	public int getTotalItemCount() {
		return this.totalItemCount;
	}
}

File: experimental/jprelude-webui/src/main/scala/org/jprelude/experimental/webui/data/PagingPosition.scala
package org.jprelude.experimental.webui.data

sealed trait PagingPosition

object PagingPosition {
  case class Undefined() extends PagingPosition
  
  case class Position(val pageIdx: Int, val totalItemCount: Int, val pageSize: Int) extends PagingPosition {
    require(pageIdx >= 0)
    require(pageSize > 0)
    require(totalItemCount > pageSize * pageIdx)
    
    def totalPageCount = ((totalItemCount - 1) / pageSize) + 1
    def offset = pageIdx * pageSize
    def isFirstPage = this.pageIdx == 0
    def isLastPage = this.pageIdx == this.totalPageCount - 1
  }
}
